my/core/query.py

```diff
@@ -0,0 +1,386 @@
+
+    reverse:        reverse the order of the resulting iterable
+
+    limit:          limit the results to this many items
+
+    drop_errors:    ignore any errors from the src
+
+    raise_errors:   raise errors when recieved from the input src
+    """
+
+    it: Iterable[ET] = []  # default
+    # check if this is a locator
+    if type(src) == tuple and len(src) == 2:  # type: ignore[arg-type]
+        it = locate_function(src[0], src[1])()  # type: ignore[index]
+    elif callable(src):
+        # hopefully this returns an iterable and not something that causes a bunch of lag when its called?
+        # should typically not be the common case, but giving the option to
+        # provide a function as input anyways
+        it = src()
+    else:
+        # assume it is already an iterable
+        if not isinstance(src, Iterable):
+            low(f"""Input was neither a locator for a function, or a function itself.
+Expected 'src' to be an Iterable, but found {type(src).__name__}...
+Will attempt to call iter() on the value""")
+        it = src
+
+    # try/catch an explicit iter() call to making this an Iterator,
+    # to validate the input as something other helpers here can work with,
+    # else raise a QueryException
+    try:
+        itr: Iterator[ET] = iter(it)
+    except TypeError as t:
+        raise QueryException("Could not convert input src to an Iterator: " + str(t))
+
+    # if both drop_errors and raise_errors are provided for some reason,
+    # should raise errors before dropping them
+    if raise_errors:
+        itr = _raise_errors(itr)
+
+    if drop_errors:
+        itr = _drop_errors(itr)
+
+    if where is not None:
+        itr = filter(where, itr)
+
+    if order_by is not None or order_key is not None or order_value is not None:
+        # we have some sort of input that specifies we should reorder the iterator
+
+        order_by_chosen: Optional[OrderFunc] = order_by  # if the user just supplied a function themselves
+        if order_by is None:
+            # https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.spy
+            [first_item], itrc = more_itertools.spy(itr)
+            # replace the 'itr' in the higher scope with itrc -- itr is consumed by more_itertools.spy
+            itr = itrc
+            # try to use a key, if it was supplied
+            # order_key doesn't use local state - it just tries to find the passed
+            # attribute, or default to the 'default' value. As mentioned above,
+            # best used for items with a similar structure
+            if order_key is not None:
+                order_by_chosen = _generate_order_by_func(first_item, key=order_key, default=default)
+                if order_by_chosen is None:
+                    raise QueryException(f"Error while ordering: could not find {order_key} on {first_item}")
+            elif order_value is not None:
+                itr1, itr2 = itertools.tee(itr, 2)  # expensive!!!
```

Comment:
I guess in principle this doesn't matter that much since `sorted` later would need linear memory anyway.
And sorting also will have at least linear slowdown, so the linear time spent computing `order_by_lookup` also isn't a big problem

Reply:
Oh also -- just curious for my understanding, not super important implementation wise I guess. In principle this could be lazy right? In the sense that the `order_by_lookup` cache could be computed for the item the first time it's used by the code below?

Reply:
I had tried to wrap my head around how to do that - the limiting factor made me write it this way was that `sorted` needed to be passed a function, and I couldn't figure out a nice way to make it lazy, while having to pass it an flags/other info on how to sort items...

It just became too much a mess of dynamically creating additional named-functions (i.e. `def some_function`) to handle state

In principle it could be, I believe, I wasn't able to figure out a nice way to do so